fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/CustomIngredientSync.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.recipe.ingredient;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import io.netty.channel.ChannelHandler;
+
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.network.PacketEncoder;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.api.ModInitializer;
+import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
+import net.fabricmc.fabric.api.networking.v1.ServerLoginConnectionEvents;
+import net.fabricmc.fabric.api.networking.v1.ServerLoginNetworking;
+import net.fabricmc.fabric.mixin.recipe.ingredient.PacketEncoderMixin;
+
+/**
+ * To reasonably support server-side only custom ingredients, we only send custom ingredients to clients that support them.
+ * If a specific client doesn't support a custom ingredient, we send the matching stacks as a regular ingredient.
+ * This is fine since all recipe computation happens server-side anyway.
+ *
+ * <p><ul>
+ *     <li>Each client sends a packet with the set of custom ingredients it supports.</li>
+ *     <li>We store that set inside the {@link PacketEncoder} using {@link PacketEncoderMixin}.</li>
+ *     <li>When serializing a custom ingredient, we get access to the current {@link PacketEncoder},
+ *     and based on that we decide whether to send the custom ingredient, or a vanilla ingredient with the matching stacks.</li>
+ * </ul>
+ */
+public class CustomIngredientSync implements ModInitializer {
+	public static final Identifier PACKET_ID = new Identifier("fabric", "custom_ingredient_sync");
+	public static final int PROTOCOL_VERSION = 1;
```

Comment:
We should rather have the server send the max supported protocol id, and the client answer back with the protocol it chooses. (e.g. in the future if we add protocol version 2 the server will send 1 or 2 depending on whether it supports it, and the client will reply with the highest protocol version that itself and the server support).

Reply:
Done